tensorflow/compiler/mlir/tensorflow/tests/convert-tf-control-flow-to-scf.mlir

```diff
@@ -0,0 +1,73 @@
+func private @testIfThen1(tensor<4xf32>) -> tensor<4xf32>
+func private @testIfElse1(tensor<4xf32>) -> tensor<4xf32>
+// CHECK-LABEL: func @testSupportedLoweringOfTFIfRegion1
+// CHECK-SAME: (%[[ARG0:.*]]: tensor<i1>, %[[ARG1:.*]]: tensor<4xf32>)
+func @testSupportedLoweringOfTFIfRegion1(%arg0: tensor<i1>, %arg1: tensor<4xf32>) -> (tensor<4xf32>, tensor<4xf32>) {
+  %res:2 = "tf.IfRegion"(%arg0) ( {
+    %call = call @testIfThen1(%arg1) : (tensor<4xf32>) -> tensor<4xf32>
+    %add = "tf.AddV2"(%call, %call) : (tensor<4xf32>, tensor<4xf32>) -> tensor<4xf32>
+    "tf.Yield"(%call, %add) : (tensor<4xf32>, tensor<4xf32>) -> ()
+  },  {
+    %call_0 = call @testIfElse1(%arg1) : (tensor<4xf32>) -> tensor<4xf32>
+    "tf.Yield"(%call_0, %call_0) : (tensor<4xf32>, tensor<4xf32>) -> ()
+  }) {is_stateless = false} : (tensor<i1>) -> (tensor<4xf32>, tensor<4xf32>)
+  return %res#0, %res#1 : tensor<4xf32>, tensor<4xf32>
+
+  // CHECK-NEXT: %[[COND:.*]] = tensor.extract %[[ARG0]][] : tensor<i1>
+  // CHECK-NEXT: %[[RES:.*]]:2 = scf.if %[[COND]] -> (tensor<4xf32>, tensor<4xf32>) {
+  // CHECK-NEXT:   %[[CALL:.*]] = call @testIfThen1(%[[ARG1]]) : (tensor<4xf32>) -> tensor<4xf32>
+  // CHECK-NEXT:   %[[ADD:.*]] = "tf.AddV2"(%[[CALL]], %[[CALL]]) : (tensor<4xf32>, tensor<4xf32>) -> tensor<4xf32>
+  // CHECK-NEXT:   scf.yield %[[CALL]], %[[ADD]] : tensor<4xf32>, tensor<4xf32>
+  // CHECK-NEXT: } else {
+  // CHECK-NEXT:   %[[CALL_0:.*]] = call @testIfElse1(%[[ARG1]]) : (tensor<4xf32>) -> tensor<4xf32>
+  // CHECK-NEXT:   scf.yield %[[CALL_0]], %[[CALL_0]] : tensor<4xf32>, tensor<4xf32>
+  // CHECK-NEXT: }
+  // CHECK-NEXT: return %[[RES]]#0, %[[RES]]#1 : tensor<4xf32>, tensor<4xf32>
+}
+
+// `tf.IfRegion` which doesn't return values gets converted to `scf.if` because its return type(s) match the type(s) of the operand(s) of the `then` and `else` terminators.
+func private @testIfThen2(tensor<4xf32>) -> ()
+func private @testIfElse2(tensor<4xf32>) -> ()
+// CHECK-LABEL: func @testSupportedLoweringOfTFIfRegion2
+// CHECK-SAME: (%[[ARG0:.*]]: tensor<i1>, %[[ARG1:.*]]: tensor<4xf32>)
+func @testSupportedLoweringOfTFIfRegion2(%arg0: tensor<i1>, %arg1: tensor<4xf32>) -> () {
+  "tf.IfRegion"(%arg0) ( {
+    call @testIfThen2(%arg1) : (tensor<4xf32>) -> ()
+    "tf.Yield"() : () -> ()
+  },  {
+    call @testIfElse2(%arg1) : (tensor<4xf32>) -> ()
+    "tf.Yield"() : () -> ()
+  }) {is_stateless = false} : (tensor<i1>) -> ()
+  return
+
+  // CHECK-NEXT: %[[COND:.*]] = tensor.extract %[[ARG0]][] : tensor<i1>
+  // CHECK-NEXT: scf.if %[[COND]] {
+  // CHECK-NEXT:   call @testIfThen2(%[[ARG1]]) : (tensor<4xf32>) -> ()
+  // CHECK-NEXT: } else {
+  // CHECK-NEXT:   call @testIfElse2(%[[ARG1]]) : (tensor<4xf32>) -> ()
+  // CHECK-NEXT: }
+  // CHECK-NEXT: return
+}
+
+// `tf.IfRegion` doesn't get converted to `scf.if` because its return type(s) don't match the type(s) of the operand(s) of the `then` and `else` terminators.
```

Comment:
Have you considered introducing a cast before the yield during the conversion to make the conversion work in all cases?

Reply:
I'll do that. Thanks for the suggestion!

Reply:
@joker-eph This is done now. The conversion happens in all cases.